stage0/src/msr.rs

```diff
@@ -0,0 +1,443 @@
+        // Preserve values of reserved bits.
+        let (old_flags, _old_memory_type) = Self::read();
+        let reserved = old_flags.bits() & !MTRRDefTypeFlags::all().bits();
+        let new_value = reserved | flags.bits() | (default_type as u64);
+        let mut msr = Self::MSR;
+        msr.write(new_value);
+    }
+}
+
+/// The x2APIC_ID register.
+///
+/// Contains the 32-bit local x2APIC ID. It is assigned by hardware at reset time, and the exact
+/// structure is manufacturer-dependent.
+///
+/// See Section 16.12 (x2APIC_ID) in the AMD64 Architecture Programmer's Manual, Volume 2 for
+/// more details.
+pub struct X2ApicIdRegister;
+
+impl X2ApicIdRegister {
+    const MSR: Msr = Msr::new(0x0000_00802);
+
+    pub unsafe fn apic_id() -> u32 {
+        (Self::MSR.read() & 0xFFFF_FFFF) as u32
+    }
+}
+
+/// x2APIC APIC Version Register
+pub struct X2ApicVersionRegister;
+
+impl X2ApicVersionRegister {
+    const MSR: Msr = Msr::new(0x0000_0803);
+}
+
+impl X2ApicVersionRegister {
+    pub unsafe fn read() -> (bool, u8, u8) {
+        let val = Self::MSR.read();
+
+        (
+            val & (1 << 31) > 0,            // EAS
+            ((val & 0xFF0000) >> 16) as u8, // MLE
+            (val & 0xFF) as u8,             // VER
+        )
+    }
+}
+
+bitflags! {
+    /// Flags in the Spurious Interrupt Register (offset 0x0F0)
+    ///
+    /// See Section 16.4.7, Spurious Interrupts, in the AMD64 Architecture Programmer's Manual, Volume 2 for more details.
+    #[derive(Clone, Copy, Debug)]
+    pub struct SpuriousInterruptFlags: u32 {
+        /// APIC Software Enable
+        const ASE = (1 << 8);
+
+        /// Focus CPU Core Checking
+        const FCC = (1 << 9);
+    }
+}
+
+/// x2APIC Spurious Interrupt Register
+pub struct X2ApicSpuriousInterruptRegister;
+
+impl X2ApicSpuriousInterruptRegister {
+    const MSR: Msr = Msr::new(0x0000_080F);
+}
+
+impl X2ApicSpuriousInterruptRegister {
+    pub unsafe fn read() -> (SpuriousInterruptFlags, u8) {
+        let val = Self::MSR.read();
+
+        (
+            SpuriousInterruptFlags::from_bits_truncate((val & 0xFFFF_FF00) as u32),
+            (val & 0xFF) as u8,
+        )
+    }
+
+    pub unsafe fn write(flags: SpuriousInterruptFlags, vec: u8) {
+        // Safety: we've estabished we're using x2APIC, so accessing the MSR is safe.
+        let val = flags.bits() as u64 | vec as u64;
+        let mut msr = Self::MSR;
+        unsafe { msr.write(val) };
+    }
+}
+
+bitflags! {
+    /// Flags in the APIC Error Status Register (offset 0x280)
+    ///
+    /// See Section 16.4.6, APIC Error Interrupts, in the AMD64 Architecture Programmer's Manual, Volume 2 for more details.
+    #[derive(Clone, Copy, Debug)]
+    pub struct ApicErrorFlags: u32 {
+        /// Sent Accept Error
+        ///
+        /// Message sent by the local APIC was not accepted by any other APIC.
+        const SAE = (1 << 2);
+
+        /// Receive Accept Error
```

Comment:
```suggestion
        /// Received Accept Error
```

Reply:
Nope, it's Receive Accept Error.

To quote chapter and verse: Figure 16-16, Chapter 16.4.6 (APIC Error Interrupts), AMD64 Architecture Programmer's Manual, Volume 2.